### PR TITLE
feat:Make use of coloured output

### PR DESCRIPTION
**Pull request by Pavankumar07s**
Task 1 completed:Create a first PR that introduce the <unistd.h> approach on a helper in utils.h/utils.cc and maybe just uses it to color code "JSON Schema CLI vX.Y.Z" on the help page. Something super simple and without letting users control it yet.
Screenshot from 2024-11-08 21-36-05

### CODE REVIEW

File: src/messages.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#ifndef MESSAGES_H
```
Comment (jviotti) [inline src/messages.h]: Please do if guards in a consistent manner, like the other headers. For example, `SOURCEMETA_JSONSCHEMA_CLI_MESSAGES_H_`

File: src/main.cc [hunk 2]
```diff
@@ -8,9 +8,16 @@
 #include <string>      // std::string
 #include <string_view> // std::string_view
 #include <vector>      // std::vector
+#ifdef _WIN32
+#include <io.h>
+#define isatty _isatty
+#else
```
Comment (jviotti) [inline src/main.cc:14]: Where are we actually checking for TTY in the code?

File: src/main.cc [hunk 3]
```diff
@@ -85,101 +92,157 @@ Global Options:
 
 For more documentation, visit https://github.com/sourcemeta/jsonschema
 )EOF"};
+// Initialize color usage
+bool use_colors = true;
```
Comment (jviotti) [inline src/main.cc]: We shouldn't have a boolean like this. Instead, each command should check the command line arguments + the TTY check.

File: src/main.cc [hunk 4]
```diff
@@ -85,101 +92,157 @@ Global Options:
...
 
 auto jsonschema_main(const std::string &program, const std::string &command,
-                     const std::span<const std::string> &arguments) -> int {
-  if (command == "fmt") {
+                     const std::span<const std::string> &arguments) -> int
```
Comment (jviotti) [inline src/main.cc]: All of these changes seem unrelated? They will also break the build, and we check formatting against `ClangFormat`

File: src/main.cc [hunk 5]
```diff
@@ -85,101 +92,157 @@ Global Options:
...
+
+    print_success(std::string(USAGE_DETAILS));
+
     return EXIT_SUCCESS;
   }
 }
 
-auto main(int argc, char *argv[]) noexcept -> int {
-  try {
+auto main(int argc, char *argv[]) noexcept -> int
```
Comment (jviotti) [inline src/main.cc]: Most of these seem to be unrelated formatting changes?

File: src/messages.h [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+#define MESSAGES_H
+
+#include <iostream>
+#include <string>
+#include "../include/termcolor/termcolor.hpp"
+
+// Declare color control as an external variable
+extern bool use_colors;
```
Comment (jviotti) [inline src/messages.h]: We should avoid these kind of global variables, even more `extern` ones.

File: src/messages.cc
Comment (jviotti) [inline src/messages.cc:1]: Mainly for this PR, I don't think we need these extra helpers. Can you use `termcolor` on standard streams directly?

File: include/termcolor/termcolor.hpp
Comment (jviotti) [inline include/termcolor/termcolor.hpp:1]: You shouldn't be including untracked external dependencies. Please add them using `DEPENDENCIES` and https://github.com/sourcemeta/vendorpull

Comment (jviotti) [reply include/termcolor/termcolor.hpp:1]: And once you add it like that, you will need to pull the dependency using CMake's `find_package`, like we do for other dependencies

**Review by jviotti — CHANGES_REQUESTED** [review]
Left a few comments

File: src/main.cc [hunk 7]
```diff
@@ -85,101 +92,157 @@ Global Options:
...
-              << " <command> [arguments...]\n";
-    std::cout << USAGE_DETAILS;
+  }
+  else
+  {
+    print_success(std::string("JSON Schema CLI - v") + std::string(sourcemeta::jsonschema::cli::PROJECT_VERSION));
+
+    print_success("Usage: " + std::filesystem::path{program}.filename().string() + " <command> [arguments...]");
+
+    print_success(std::string(USAGE_DETAILS));
```
Comment (jviotti) [inline src/main.cc]: We probably shouldn't colorise the entire thing. Maybe just colorise the usage line above or something like that?

File: src/main.cc (hunk 2 above)
Comment (Pavankumar07s) [reply src/main.cc:14]: Yes

Comment (Pavankumar07s) [reply src/main.cc:14]: We haven't introduced yet . In this pr i have just inlmported the module for tty and color. In next pr we will introduce condition for tty as cli-program detected then we will continue using colors else we will turn off.

File: src/main.cc (hunk 3 above)
Comment (Pavankumar07s) [reply src/main.cc]: Alright ,in our code i was just going to check ones when the problem starts.Thankyou for guidance i will make sure it checks each time when command run.

File: src/main.cc (hunk 4 above)
Comment (Pavankumar07s) [reply src/main.cc]: Thankyou ,i was also confused i will now use formated code for our simplicity and build.

File: src/main.cc (hunk 5 above)
Comment (Pavankumar07s) [reply src/main.cc]: I am going to revert al these formatting. Thankyou.

File: src/messages.h (hunk 6 above)
Comment (Pavankumar07s) [reply src/messages.h]: Alright @jviotti 👍

File: src/messages.cc
Comment (Pavankumar07s) [reply src/messages.cc:1]: Yes we can do but in further steps ,we again have to change .Is is better to define all colors in one messages page so that it would be easy and maintained. Pls share your thoughts.

File: src/main.cc (hunk 7 above)
Comment (Pavankumar07s) [reply src/main.cc]: Lets just color the usage part.

File: include/termcolor/termcolor.hpp
Comment (Pavankumar07s) [reply include/termcolor/termcolor.hpp:1]: yeaah i got it .Thankyou.

File: src/messages.cc
Comment (jviotti) [reply src/messages.cc:1]: I think `termcolor` interface is already quite concise and we could use it directly when needed.

File: src/main.cc (hunk 4 above)
Comment (jviotti) [reply src/main.cc]: If you run `make` on your system that should run `ClangFormat` as per the project's configuration and always get the formatting right

File: src/main.cc (hunk 2 above)
Comment (jviotti) [reply src/main.cc:14]: I think you still need the TTY check on this one, otherwise i.e. piping the help text will result in escaped characters

File: src/messages.cc
Comment (Pavankumar07s) [reply src/messages.cc:1]: Alright @jviotti 👍

File: src/main.cc (hunk 4 above)
Comment (Pavankumar07s) [reply src/main.cc]: Yeah i am this one only Thankyou you are great help.😃

File: src/main.cc (hunk 2 above)
Comment (Pavankumar07s) [reply src/main.cc:14]: Alright @jviotti 👍

**Comment by Pavankumar07s** [comment]
Hiii @jviotti whats i am doing wrong ? When i am using make command it is showing 100% test passes but the workflow is breaking...can you guide me.
Screenshot from 2024-11-13 13-13-32
Screenshot from 2024-11-13 12-54-46
Screenshot from 2024-11-13 12-55-05
